### PR TITLE
apidoc check and improvement

### PR DESCRIPTION
**Pull request by ebarboni**
This PR is a modification of apidoc generation to prepare the cleanup #4450

Compared to #4450, only do a full verification if parameter apidocfullcheck is set to true. 

Disallowedlinks is anticipated fixed #4450 generate txt file to list rejected/accepted to later processing (linkchecker)
Fixed xsl to new website.
linelength 240
check all html file not only arch.
anchor regexp should also scan for id
build.properties revised to match level stability and also master/release branding dependencies

Javadoc indexer changed to json and parse more information to get interface #4699

This PR should be put in each release branch later.

Hope this is shorter to review :D and so next PR

### CODE REVIEW

**Review by mbien — APPROVED** [review]
couldn't find any obvious issues :)

File: nbbuild/antsrc/org/netbeans/nbbuild/CheckLinks.java [hunk 1]
```diff
@@ -633,8 +645,33 @@ final Boolean isOk (URI u) throws BuildException {
                 throw new BuildException ("Each filter must have pattern attribute");
             }
             
-            if (pattern.matcher (u.toString ()).matches ()) {
-                log ("Matched " + u + " accepted: " + accept, org.apache.tools.ant.Project.MSG_VERBOSE);
+            if (pattern.matcher(u.toString()).matches()) {
+                log("Matched " + u + " accepted: " + accept, org.apache.tools.ant.Project.MSG_VERBOSE);
+                if (externallinksdump != null) {
+                    try {
+                        String dummyName;
```
Comment (lbownik) [inline nbbuild/antsrc/org/netbeans/nbbuild/CheckLinks.java:652]: Wouldn't ternary operator be more readable here?

File: nbbuild/antsrc/org/netbeans/nbbuild/CheckLinks.java [hunk 2]
```diff
@@ -633,8 +645,33 @@ final Boolean isOk (URI u) throws BuildException {
...
+                            dummyName = "acceptednetbeans.txt";
+                        } else {
+                            dummyName = "rejectednetbeans.txt";
+                        }
+                        Path dumppath = externallinksdump.toPath().resolve(dummyName);
+                        if (Files.notExists(dumppath)) {
+                            Files.createDirectories(dumppath.getParent());
+                            Files.createFile(dumppath);
+                        }
+                        List<String> existingEntries = Files.readAllLines(externallinksdump.toPath().resolve(dummyName));
```
Comment (lbownik) [inline nbbuild/antsrc/org/netbeans/nbbuild/CheckLinks.java:663]: existing entries variable seems unneeded.
would
Set<String> sortedEntries = new TreeSet<>(Files.readAllLines(externallinksdump.toPath().resolve(dummyName)));
rork?

File: nbbuild/antsrc/org/netbeans/nbbuild/CheckLinks.java [hunk 3]
```diff
@@ -633,8 +645,33 @@ final Boolean isOk (URI u) throws BuildException {
...
+                        Set<String> sortedEntries = new TreeSet<>(existingEntries);
+                        sortedEntries.add(u.toString());
+                        StringBuilder sb = new StringBuilder();
+                        for (String sortedEntry : sortedEntries) {
+                            sb.append(sortedEntry).append(System.lineSeparator());
+                        }
+                        Files.write(dumppath, sb.toString().getBytes(), StandardOpenOption.WRITE, StandardOpenOption.TRUNCATE_EXISTING);
```
Comment (lbownik) [inline nbbuild/antsrc/org/netbeans/nbbuild/CheckLinks.java:670]: Wouldn't 
Files.write(dumbpath, sortedEntries, StandardOpenOption.WRITE, StandardOpenOption.TRUNCATE_EXISTING);
be more efficient?

File: nbbuild/antsrc/org/netbeans/nbbuild/JavadocIndex.java [hunk 4]
```diff
@@ -85,90 +91,72 @@ public void execute() throws BuildException {
...
+                            className,
+                            "interface".equals(title[0].trim()),
+                            urlPrefix + "/" + anchor.attr("href")
                     );
-                    if (c.name == null) throw new NullPointerException ("Null name for " + line + "\nclass: " + c);
-                    if (c.name.length () == 0) throw new IllegalStateException ("Empty name for " + line + "\nclass: " + c);
-                    
-                    log ("Adding class: " + c, Project.MSG_DEBUG);
-                    
+                    if (c.name.length() == 0) {
```
Comment (lbownik) [inline nbbuild/antsrc/org/netbeans/nbbuild/JavadocIndex.java:139]: how about 
if(c.name.isEmpty()) {

File: nbbuild/antsrc/org/netbeans/nbbuild/CheckLinks.java (hunk 2 above)
Comment (ebarboni) [reply nbbuild/antsrc/org/netbeans/nbbuild/CheckLinks.java:663]: will change that thanks

File: nbbuild/antsrc/org/netbeans/nbbuild/CheckLinks.java (hunk 3 above)
Comment (ebarboni) [reply nbbuild/antsrc/org/netbeans/nbbuild/CheckLinks.java]: +1

**Pull request merged**